### PR TITLE
feat: add tailwind + shadcn/ui.

### PR DESCRIPTION
**Pull request by Emt-lin**
Support out-of-the-box the Tailwind + shadcn/ui functionality

### CODE REVIEW

File: styles.css
Comment (zeroliu) [inline styles.css]: should this file be checked in? It looks like the generated tailwind CSS file

File: src/styles/tailwind.css
Comment (zeroliu) [inline src/styles/tailwind.css]: what are these styles?

File: src/styles/tailwind.css
Comment (zeroliu) [inline src/styles/tailwind.css]: to make UI library compatible with obsidian plugin, we should use the CSS variables defined in obsidian (https://docs.obsidian.md/Reference/CSS+variables/CSS+variables). Many obsidian users use themes, which are made by altering those CSS variables (https://docs.obsidian.md/Themes/App+themes/Build+a+theme). Users would expect a well-designed plugin to change the appearance accordingly.

This is the main reason why I believe shadcn is the right choice for obsidian plugin UI library. It provides developer great flexibility to change how components are styled. However, we will lose the value if we use the shadcn default CSS variables.

I suggest:
1. Do not introduce new CSS variables
2. update tailwind.config.js to create tokens that map to obsidian CSS variables, including color, spacing, sizing, etc
3. whenever we introduce a shadcn component, restyle it with the new token before we integrate it into the plugin.

**Comment by logancyang** [comment]
I checked out this branch and ran locally, it seems to override some obsidian native CSS vars
SCR-20241203-seis

**Comment by Emt-lin** [comment]
@logancyang @zeroliu 

now, `style.css` is already generated by tailwind cli. The original traditional css style are in`tailwind.css` file. Perhaps We should make these traditional css style into tailwind style.

**Comment by logancyang** [comment]
> @logancyang @zeroliu
> 
> now, `style.css` is already generated by tailwind cli. The original traditional css style are in`tailwind.css` file. Perhaps We should make these traditional css style into tailwind style.

Is this going to be quite a big effort to make everything look the same/similar? I'm not familiar with how it can be converted though. @zeroliu is the expert here I'll wait for his recommendation.

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: @zeroliu sometimes, I don't want to use some of the built-in theme styles.`(eg: radius)`
so, I think we should select some css styles to map.

**Comment by Emt-lin** [comment]
> > @logancyang @zeroliu
> > now, `style.css` is already generated by tailwind cli. The original traditional css style are in`tailwind.css` file. Perhaps We should make these traditional css style into tailwind style.
> 
> Is this going to be quite a big effort to make everything look the same/similar? I'm not familiar with how it can be converted though. @zeroliu is the expert here I'll wait for his recommendation.

I already delete `styles.css` file in latest commit, it will be created by `tailiwind cli` When you are running `npm run dev or build`

File: src/styles/tailwind.css
Comment (logancyang) [inline src/styles/tailwind.css]: These 3 red blocks were added recently, why are they removed here?

File: package.json
Comment (logancyang) [inline package.json]: Typo: `build:tailiwind`. I recommend some form of AI assistance to spot these things in the future. It's hard for reviewers to spot.

File: package.json
Comment (logancyang) [inline package.json]: How about something like
```
{
  "scripts": {
    "build:tailwind": "npx tailwindcss -i src/styles/tailwind.css -o ./styles.css --minify",
    "build:esbuild": "tsc -noEmit -skipLibCheck && node esbuild.config.mjs production",
    "build": "npm run build:tailwind && npm run build:esbuild || exit 1"
  }
}
```

**Comment by logancyang** [comment]
Another concern is that the bundle size is up significantly from 5MB+ to 21MB+. Is there any way we can make it smaller?

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: @logancyang Maybe I don't use the latest `styles.css` in master branch. I will replace it by  the lasted `styles.css`

File: src/styles/app.css
Comment (zeroliu) [inline src/styles/app.css:1]: what is this file?

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js]: can we not extend but redefine all the tokens? Otherwise, people can still use predefined tailwind tokens without realizing that it's not connected to obsidian CSS variables. We can also turn on this eslint rule to enforce only supported classnames are used in the code base: https://github.com/francoismassart/eslint-plugin-tailwindcss/blob/master/docs/rules/no-custom-classname.md

File: src/components/ui/input.tsx
Comment (zeroliu) [inline src/components/ui/input.tsx]: do we support file type input? If not, I suggest not adding the related style until it's needed

File: src/components/ui/input.tsx
Comment (zeroliu) [inline src/components/ui/input.tsx]: not blocking this PR but maybe we can look into React 19 and use the `ref` props now that it's officially GA. Maintaining forwardRef for RadixUI can be pretty tedious.

**Review by zeroliu — COMMENTED** [review]
Left some more comments. Looking good overall! Looking forward to using tailwindCSS in the near future

**Comment by logancyang** [comment]
@Emt-lin Thanks for the sync today! When you get a chance, could you address zero's comments? And I think this should be ready to merge after!

File: src/styles/app.css
Comment (Emt-lin) [reply src/styles/app.css:1]: nothing, I will remove it.

File: src/components/ui/input.tsx
Comment (Emt-lin) [reply src/components/ui/input.tsx]: yeah, I agree with your opinion. However, I don't want to increase the complexity of maintenance If I remove the styles related to file type input.

File: src/constants.ts
Comment (zeroliu) [inline src/constants.ts:127]: does it belong to this PR?

File: src/styles/tailwind.css
Comment (zeroliu) [inline src/styles/tailwind.css]: why do we need to redefine obsidian CSS with `--ob-bg` prefix?

File: src/styles/tailwind.css
Comment (zeroliu) [inline src/styles/tailwind.css]: I don't think we should be touching the default component style as it will affect UI outside of obsidian-copilot

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js:7]: we can use tailwind.config.ts to get full type support instead of this type comment

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js]: what's the use case of black, white, red, yellow? Are there existing semantic tokens that can replace them?

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js:152]: When do we need this token? It doesn't feel necessary to introduce a new token that needs to calculate a new value outside of the obsidian design system?

File: src/settings/model.ts
Comment (zeroliu) [inline src/settings/model.ts:1]: does this change belong to this PR?

File: src/constants.ts
Comment (Emt-lin) [reply src/constants.ts:127]: yeah,  it's a separate commit and @logancyang also learn about this.

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: I think that  redefine obsidian CSS will more clarify. I want to provide out-of-the-box functionality, when we add some shadcn/ui component.

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: @zeroliu It doesn't affect UI outside of obsidian-copilot, It only work for copilot ui.

File: tailwind.config.js
Comment (Emt-lin) [reply tailwind.config.js:7]: The JS type documentation also provides full type support. Therefore, both TS and JS can work.

File: tailwind.config.js
Comment (Emt-lin) [reply tailwind.config.js]: I think we don't need full semantic tokens support. It only base color. We don't need to fully trap ourselves in the concept of Obsidian. It's also tailwind project.

File: tailwind.config.js
Comment (Emt-lin) [reply tailwind.config.js:152]: In the button component, I need the define of `5.5`, others else can be removed.

File: src/styles/tailwind.css
Comment (zeroliu) [reply src/styles/tailwind.css]: Do you have any specific features in mind that would require us to break apart from the default css variables? This is an additional layer that we have to maintain. If we want to apply a new obsidian css variable that is not included in this PR, the process of adding it would require us to:
1. Make a copy of the css variable 
2. Create a new tailwind token with a different semantic value 

The more undocumented overhead we add, the more difficult it is for new contributors to make changes. 

If you feel strongly that creating our own semantic token is necessary, and it can simplify the component library adoption, I'm supportive of that. However, we should at least consider not adding another layer of complexity and use the obsidian css variable directly.

Also worth noting that any external CSS variable that we introduce will not be compatible with the obsidian theme, which can be a dealbreaker for some users.

File: tailwind.config.js
Comment (zeroliu) [reply tailwind.config.js]: Using colors like white and black directly will likely to cause dark theme to break, as well as causing unexpected edge cases for people using non default themes. Do you have a good way to avoid these potential breakage?

File: src/styles/tailwind.css
Comment (zeroliu) [reply src/styles/tailwind.css]: This will affect all matching components. I created a sample plugin and added a button with "select-trigger-reset" class. By loading this version of copilot, my settings button now looks different because of this style set in copilot plugin:

Screenshot 2024-12-21 at 9 04 53 PM

Comment (zeroliu) [reply src/styles/tailwind.css]: Just checked again, this is not a new problem introduced by this PR. The existing style can also affect other plugins. If I add `warning-message` to my plugin element and load the master version copilot, my element will also be affected. I sent a question to the obsidian discord to understand how style isolation works in obsidian. It seems very error-prone.

File: tailwind.config.js
Comment (Emt-lin) [reply tailwind.config.js]: I test some top theme. It's work well in light/dark mode. Dialog overlay will use the black color. I think you can pull up pr #955, you will know about something.

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: `select-trigger-reset` is my custom css classname. Others have a very low probability of using it. 

> This will affect all matching components. 

I was aware of css conflicts when developing the UI earlier.

The reason：
CleanShot 2024-12-23 at 10 42 55

File: src/styles/tailwind.css
Comment (Emt-lin) [reply src/styles/tailwind.css]: I am merely listing all the obsidian css variables rather than defining a new set of variables.

> If we want to apply a new obsidian css variable

I have listed most of the color css variables.

> The more undocumented overhead we add, the more difficult it is for new contributors to make changes.

When a new contributor joins, he/her doesn't need to do anything extra work, and he/her doesn't need to care about these. 

He can use existing components or download new ones from the shadcn/ui, and then he doesn't need to do anything, the component will automatically change with the theme.

**Comment by Emt-lin** [comment]
@logancyang @zeroliu please review this pr.

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js]: please clean up the comments

**Review by zeroliu — APPROVED** [review]

File: tailwind.config.js
Comment (zeroliu) [inline tailwind.config.js]: can we not use the obsidian defined red and yellow color?

Comment (Emt-lin) [reply tailwind.config.js]: The defalut color offers more choices. Why we need the red color?

CleanShot 2024-12-26 at 14 21 29

Comment (zeroliu) [reply tailwind.config.js]: This is the principle of limiting the choice for developers and consistency. 

Regarding limiting the choice, TailwindCSS is a generic CSS library; its goal is to allow people to build any applications. TW has to provide not only a complete set of primitive tokens and semantic tokens and let people decide what they want. Our goal is not to replicate TailwindCSS tokens but to tailor them down to what the copilot plugin may need. We don't want developers to add tokens like `bg-color-red-50` without a specific reason in the plugin because #fef2f2 is not a color in either the obsidian design system or copilot design.

For consistency, this PR defines red, yellow, and black differently from the rest of the colors, making it harder to predict when applying color tokens. We would either go with Tailwind color tokens entirely or choose only obsidian colors. Staying in the middle will make decision-making harder in the future.

Unless we have a concrete use case that the obsidian color tokens cannot fit and it adds value to the copilot design, we should refrain from introducing new tokens.

Comment (Emt-lin) [reply tailwind.config.js]: > Staying in the middle will make decision-making harder in the future.

agree.

But Why we need red color? Isn't it a primitive color? It's not a semantic token.

Comment (Emt-lin) [reply tailwind.config.js]: below pic -> text: yellow-700, bg: yellow-50, What should I do? eg: text-yellow-rgb/10, bg-yellow-rgb/50?

Google Chrome 2024-12-26 16 09 28

Comment (zeroliu) [reply tailwind.config.js]: This is the callout component used in the obsidian app and site. 
image


It uses --callout-warning which is --color-orange-rgb. The background is rgba(var(--callout-warning), 0.1)). There is no specific token for it. 

My obsidian tailwind project didn't include all the component tokens from obsidian yet. You can add this token to your PR. Then you can build it as `bg-callout-warning/10` and `text-callout-warning`. Here's the reference https://docs.obsidian.md/Reference/CSS+variables/Editor/Callout

**Comment by logancyang** [comment]
Just to confirm, this new setup compiles tailwind to `styles.css`, yes? So the release process is still to run `npm run build` and release `main.js`, `styles.css` and `manifest.json` without any additional steps, correct? @Emt-lin @zeroliu

**Comment by Emt-lin** [comment]
> Just to confirm, this new setup compiles tailwind to `styles.css`, yes? So the release process is still to run `npm run build` and release `main.js`, `styles.css` and `manifest.json` without any additional steps, correct? @Emt-lin @zeroliu

yeah. cc @zeroliu

**Pull request merged**
